bika/lims/browser/bika_listing.py

```diff
@@ -1125,48 +1019,350 @@ def folderitems(self, full_objects=False, classic=True):
                 idx += 1
         return results
 
-    def _fetch_brains(self, idxfrom=0):
-        """Returns the brains that must be displayed in the current list
-
-        Uses the contentFilter and/or contentsMethod class variables (or
-        functions) to query against the database. Also takes into account if
-        only a subset of the results must be returned by using idxfrom and. If
-        the number of results is lower than idxfrom, will return an empty array
-
-        :param idxfrom: index to start to count for results
-        :return: the list of brains to be displayed in this list
-        """
-        # Creating a copy of the contentFilter dictionary in order to include
-        # the filter bar's filtering additions in the query. We don't want to
-        # modify contentFilter with those 'extra' filtering elements to be
-        # included in the.
-        contentFilterTemp = copy.deepcopy(self.contentFilter)
-        addition = self.get_filter_bar_queryaddition()
+    def get_catalog(self, default="portal_catalog"):
+        """Get the catalog tool to be used in the listing
+
+        :returns: ZCatalog tool
+        """
+        try:
+            return api.get_tool(self.catalog)
+        except api.BikaLIMSError:
+            return api.get_tool(default)
+
+    @viewcache.memoize
+    def get_catalog_indexes(self):
+        """Return a list of registered catalog indexes
+        """
+        return self.get_catalog().indexes()
+
+    @viewcache.memoize
+    def get_columns_indexes(self):
+        """Returns a list of allowed sorting indexeds
+        """
+        columns = self.columns
+        indexes = [v["index"] for k, v in columns.items() if "index" in v]
+        return indexes
+
+    def get_sort_order(self):
+        """Get the sort_order criteria from the request or view
+        """
+        form_id = self.get_form_id()
+        key = "{}_sort_order".format(form_id)
+        sort_order = self.request.get(key, None)
+
+        if sort_order is None:
+            sort_order = self.contentFilter.get("sort_order", self.sort_order)
+        if sort_order not in ["ascending", "descending"]:
+            sort_order = "descending"
+
+        return sort_order
+
+    def is_valid_sort_index(self, sort_on):
+        """Checks if the sort_on index is capable for a sort_
+
+        :param sort_on: The name of the sort index
+        :returns: True if the sort index is capable for sorting
+        """
+        # List of known catalog indexes
+        catalog_indexes = self.get_catalog_indexes()
+        if sort_on not in catalog_indexes:
+            return False
+        catalog = self.get_catalog()
+        sort_index = catalog.Indexes.get(sort_on)
+        if not hasattr(sort_index, 'documentToKeyMap'):
```

Comment:
pheew!, didn't know that an index without this `documentToKeyMap` attr was not suitable for sorting, good catch!
I'd add something like `logger.warning("Index {} cannot be used for sorting".format(sort_on))`